File: backends/java/zappy-adapters/src/main/java/com/zappymart/adapters/persistence/JpaCustomerRepository.java
package com.zappymart.adapters.persistence;

import com.zappymart.application.ports.CustomerRepository;
import com.zappymart.domain.accounts.Customer;
import com.zappymart.domain.shared.EmailAddress;

import java.util.Optional;

public final class JpaCustomerRepository implements CustomerRepository {

    private final CustomerRowRepository customerRows;

    JpaCustomerRepository(CustomerRowRepository customerRows) {
        this.customerRows = customerRows;
    }

    @Override
    public Optional<Customer> byId(String customerId) {
        return customerRows.findById(customerId).map(JpaCustomerRepository::customerOf);
    }

    @Override
    public Optional<Customer> byEmail(EmailAddress email) {
        return customerRows.findByEmail(email.value()).map(JpaCustomerRepository::customerOf);
    }

    @Override
    public Customer save(Customer customer) {
        customerRows.save(new CustomerRow(customer.id(), customer.email().value(), customer.name(),
                customer.passwordHash(), customer.createdAt()));
        return customer;
    }

    private static Customer customerOf(CustomerRow row) {
        return new Customer(row.id, new EmailAddress(row.email), row.name, row.passwordHash, row.createdAt);
    }
}

File: backends/kotlin/zappy-adapters/src/main/kotlin/nl/zappymart/adapters/persistence/JpaCustomerRepository.kt
package nl.zappymart.adapters.persistence

import nl.zappymart.adapters.persistence.entities.CustomerEntity
import nl.zappymart.application.ports.CustomerRepository
import nl.zappymart.domain.accounts.Customer
import nl.zappymart.domain.shared.EmailAddress
import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Transactional

@Repository
@Transactional(readOnly = true)
class JpaCustomerRepository(private val customers: CustomerEntities) : CustomerRepository {

    override fun findById(customerId: String): Customer? =
        customers.findById(customerId).orElse(null)?.asCustomer()

    override fun findByEmail(email: EmailAddress): Customer? =
        customers.findByEmail(email.value)?.asCustomer()

    @Transactional
    override fun save(customer: Customer): Customer {
        val entity = customers.findById(customer.id).orElseGet { CustomerEntity(id = customer.id) }
        entity.email = customer.email.value
        entity.name = customer.name
        entity.passwordHash = customer.passwordHash.value
        entity.createdAt = customer.createdAt
        return customers.save(entity).asCustomer()
    }
}
